### PR TITLE
ddl: set column collate from ColumnOptionCollate

### PR DESCRIPTION
**Pull request by bb7133**
### What problem does this PR solve? 
In MySQL `CREATE TABLE` syntax, column collation is part of the column option, so the following statements are valid(although meaningless):

```
create table t(a decimal collate utf8_bin)
create table t(a decimal collate utf8_bin collate utf8_general_ci)
```

This PR makes those statements acceptable

### What is changed and how it works?
In TiDB parser, Column "COLLATE" definitions are moved from 'StringType' to 'ColumnOption'(https://github.com/pingcap/parser/pull/260)

This PR uses ColumnOptionCollate to set the collation of a column

### Check List

Tests
 - Integration test

Code changes
 - Has interface methods change

Side effects
 - Increased code complexity

Related changes
 - None

### CODE REVIEW

**Comment by winkyao** [comment]
Please fix ci.

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -430,6 +430,8 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef, o
 				col.GeneratedStored = v.Stored
 				_, dependColNames := findDependedColumnNames(colDef)
 				col.Dependences = dependColNames
+			case ast.ColumnOptionCollate:
```
Comment (crazycs520) [inline ddl/ddl_api.go]: I'm afraid you can't simply set the collate with out any check.
In This PR
```sql
test > create table t1(a varchar(10) collate utf8_bin, b decimal collate utf8_bin);
test > show create table t1
+-------+------------------------------------------------------------------------+
| Table | Create Table                                                           |
+-------+------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (                                                    |
|       |   `a` varchar(10) CHARACTER SET utf8mb4 COLLATE utf8_bin DEFAULT NULL, |
|       |   `b` decimal(11,0) DEFAULT NULL                                       |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin            |
+-------+------------------------------------------------------------------------+
test > show full columns from t1;
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
| Field | Type          | Collation | Null | Key | Default | Extra | Privileges                      | Comment |
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
| a     | varchar(10)   | utf8_bin  | YES  |     | <null>  |       | select,insert,update,references |         |
| b     | decimal(11,0) | utf8_bin  | YES  |     | <null>  |       | select,insert,update,references |         |
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
```

In MYSQL
```sql
test > create table t1(a varchar(10) collate utf8_bin, b decimal collate utf8_bin);
test > show create table t
(1146, u"Table 'test.t' doesn't exist")
test > show create table t1
+-------+---------------------------------------------------------------------+
| Table | Create Table                                                        |
+-------+---------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (                                                 |
|       |   `a` varchar(10) CHARACTER SET utf8 COLLATE utf8_bin DEFAULT NULL, |
|       |   `b` decimal(10,0) DEFAULT NULL                                    |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=ascii                               |
+-------+---------------------------------------------------------------------+
test > show full columns from t1
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
| Field | Type          | Collation | Null | Key | Default | Extra | Privileges                      | Comment |
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
| a     | varchar(10)   | utf8_bin  | YES  |     | <null>  |       | select,insert,update,references |         |
| b     | decimal(10,0) | <null>    | YES  |     | <null>  |       | select,insert,update,references |         |
+-------+---------------+-----------+------+-----+---------+-------+---------------------------------+---------+
```

Comment (bb7133) [reply ddl/ddl_api.go]: yeah, I was planning to put the checks in the subsequent PRs

Comment (bb7133) [reply ddl/ddl_api.go]: hi @crazycs520, the case you showed has been fixed in https://github.com/pingcap/tidb/pull/10007.

Another known issue is the correlation between Charset/Collation: https://github.com/pingcap/tidb/issues/10122, I plan to fix it in another PR to keep this one clear enough.

PTAL

Comment (crazycs520) [reply ddl/ddl_api.go]: em... I prefer to add a check here: 
```go
                if field_types.HasCharset(colDef.Tp) {
					col.FieldType.Collate = v.StrValue
				}
```

Comment (bb7133) [reply ddl/ddl_api.go]: This is different from MySQL's way, but I think it's fine to do so

Comment (bb7133) [reply ddl/ddl_api.go]: addressed, PTAL @crazycs520

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -440,6 +441,10 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef, o
...
+				if field_types.HasCharset(colDef.Tp) {
+					col.FieldType.Collate = v.StrValue
```
Comment (winkyao) [inline ddl/ddl_api.go]: Just set collate, keep the charset empty?

Comment (bb7133) [reply ddl/ddl_api.go]: well, I think maybe another PR is needed to solve the correlation between charset and collation.

**Pull request merged**